rmw_zenoh_cpp/src/detail/rmw_node_data.cpp

```diff
@@ -0,0 +1,49 @@
+// Copyright 2024 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "rmw_node_data.hpp"
+
+#include <utility>
+
+#include "logging_macros.hpp"
+
+#include "rcpputils/scope_exit.hpp"
+
+namespace rmw_zenoh_cpp
+{
+///=============================================================================
+NodeData::NodeData(
+  std::size_t id,
+  std::shared_ptr<liveliness::Entity> entity,
+  zc_owned_liveliness_token_t token)
+: id_(std::move(id)),
+  entity_(std::move(entity)),
+  token_(std::move(token))
+{
+  // Do nothing.
+}
+
+///=============================================================================
+NodeData::~NodeData()
+{
+  zc_liveliness_undeclare_token(z_move(token_));
```

Comment:
I'm not a huge fan of having different entities declare/undeclare the token.  In this case, I'd prefer if we moved the declaration of the token into the constructor; that way there is no ambiguity of which entity owns it.

Reply:
Good call out. It was this way before I moved the constructor to `public` scope. Now i've implemented a `make()` function that will return `nullptr` if construction of the entity or liveliness token fails. https://github.com/ros2/rmw_zenoh/pull/259/commits/bda3ed69541a6f886563429dcb1363ca47310752